### PR TITLE
Fix rules toggle.

### PR DESCRIPTION
**Pull request by kekek2**
If you switch the state of more than one hundred rules in IDS, they are not always switched. In parallel, run multiple AJAX requests simultaneously writing changes to the config.xml. Because of this, some changes may be lost. I made AJAX requests to be sent sequentially.

### CODE REVIEW

File: src/opnsense/mvc/app/views/OPNsense/IDS/index.volt [hunk 1]
```diff
@@ -199,19 +199,30 @@ POSSIBILITY OF SUCH DAMAGE.
                 $.each(rows, function(key,uuid){
                     keyset.push(uuid);
                     if ( combine == undefined || keyset.length > combine) {
-                        deferreds.push(ajaxCall(url + keyset.join(',') +'/'+url_suffix, sendData={},null));
+                        (function (localKeyset) {
+                            promise = promise.then(function () {
+                                return ajaxCall(url + localKeyset.join(',') + '/' + url_suffix, sendData = {}, null);
+                            })
+                        }) (keyset);
```
Comment (fabianfrz) [inline src/opnsense/mvc/app/views/OPNsense/IDS/index.volt]: this function is probably not needed

File: src/opnsense/mvc/app/views/OPNsense/IDS/index.volt [hunk 2]
```diff
@@ -199,19 +199,30 @@ POSSIBILITY OF SUCH DAMAGE.
...
                 });
 
                 // flush remaining items
                 if (keyset.length > 0) {
-                    deferreds.push(ajaxCall(url + keyset.join(',') +'/'+url_suffix, sendData={},null));
+                    (function (localKeyset) {
+                        promise = promise.then(function () {
+                            return ajaxCall(url + localKeyset.join(',') + '/' + url_suffix, sendData = {}, null);
+                        })
+                    })(keyset);
```
Comment (fabianfrz) [inline src/opnsense/mvc/app/views/OPNsense/IDS/index.volt]: same here

File: src/opnsense/mvc/app/views/OPNsense/IDS/index.volt (hunk 1 above)
Comment (kekek2) [reply src/opnsense/mvc/app/views/OPNsense/IDS/index.volt]: First I did without closure. All calls AjaxCall was with the last value of the keyset. I had to add this feature.

Comment (fabianfrz) [reply src/opnsense/mvc/app/views/OPNsense/IDS/index.volt]: I am talking about the outer function, not the promise

**Comment by kekek2** [comment]
var keyset = [];
                $.each(rows, function (key, uuid) {
                    keyset.push(uuid);
                    if (combine == undefined || keyset.length > combine) {
                        promise = promise.then(function () {
                            return ajaxCall(url + keyset.join(',') + '/' + url_suffix, sendData = {}, null);
                        })
                        keyset = [];
                    }
                });

                if (keyset.length > 0) {
                    promise = promise.then(function () {
                        return ajaxCall(url + keyset.join(',') + '/' + url_suffix, sendData = {}, null);
                    })
                }
`
These screenshots I made with this code. All requests are the same.

**Comment by kekek2** [comment]
Screenshots with my code

**Comment by fabianfrz** [comment]
looks like you are loosing the reference - should be fixable by `let` and local variables.

File: src/opnsense/mvc/app/views/OPNsense/IDS/index.volt [hunk 3]
```diff
@@ -196,25 +196,23 @@ POSSIBILITY OF SUCH DAMAGE.
...
                 if (keyset.length > 0) {
-                    (function (localKeyset) {
-                        promise = promise.then(function () {
-                            return ajaxCall(url + localKeyset.join(',') + '/' + url_suffix, sendData = {}, null);
-                        })
-                    })(keyset);
+                    const localKeyset = keyset;
+                    promise = promise.then(function () {
+                        return ajaxCall(url + localKeyset.join(',') + '/' + url_suffix, sendData = {}, null);
+                    })
```
Comment (fabianfrz) [inline src/opnsense/mvc/app/views/OPNsense/IDS/index.volt]: add a semicolon

**Pull request closed without merging**